
# Numpy array vs Python lists

In [1]:
a = [i for i in range(10000000)]
b = [i for i in range(10000000,20000000)]

c = []
import time
start = time .time()
for i in range(len(a)):
    c.append(a[i] + b[i])
print(time.time()-start)

1.4621992111206055


In [2]:
# numpy
import numpy as np
a = np.arange(10000000)
b = np.arange(10000000,20000000)

start = time.time()
c = a + b
print(time.time()-start)

0.08923482894897461


# Here we can say that numpy is faster than python lists
* 1.462 vs 0.0892

# how much faster
in my case its 16 times faster

In [6]:
1.4621992111206055/0.08923482894897461

16.385969787163553

# memory usage

# python

In [7]:
a = [i for i in range (10000000)]
import sys
sys.getsizeof(a)

89095160

In [8]:
a = np.arange(10000000, dtype = np.int8)
sys.getsizeof(a)

10000112

# convenience

### Advanced Indexing

In [9]:
# Normal Indexing and slicing

a = np.arange(24).reshape(6,4)
a

array([[ 0,  1,  2,  3],
       [ 4,  5,  6,  7],
       [ 8,  9, 10, 11],
       [12, 13, 14, 15],
       [16, 17, 18, 19],
       [20, 21, 22, 23]])

In [10]:
a[1,2]

np.int64(6)

In [12]:
a[1:3,1:3]

array([[ 5,  6],
       [ 9, 10]])

In [13]:
a

array([[ 0,  1,  2,  3],
       [ 4,  5,  6,  7],
       [ 8,  9, 10, 11],
       [12, 13, 14, 15],
       [16, 17, 18, 19],
       [20, 21, 22, 23]])

#  fancy index 
 *  jo chaiey wo pass 
 *

In [15]:
a[[0,2,3,4]]

array([[ 0,  1,  2,  3],
       [ 8,  9, 10, 11],
       [12, 13, 14, 15],
       [16, 17, 18, 19]])

In [ ]:
a= np.arange(24).reshape(6,4)


In [19]:
a
a[[0,1,4,5]]


array([[ 0,  1,  2,  3],
       [ 4,  5,  6,  7],
       [16, 17, 18, 19],
       [20, 21, 22, 23]])

In [22]:
a

array([[ 0,  1,  2,  3],
       [ 4,  5,  6,  7],
       [ 8,  9, 10, 11],
       [12, 13, 14, 15],
       [16, 17, 18, 19],
       [20, 21, 22, 23]])

In [ ]:
a[:,[0,2,3]] 

array([[ 0,  2,  3],
       [ 4,  6,  7],
       [ 8, 10, 11],
       [12, 14, 15],
       [16, 18, 19],
       [20, 22, 23]])

In [24]:
# Boolean Indexing
a = np.random.randint(1,100,24).reshape(6,4)
a

array([[93,  4, 65, 51],
       [17, 26, 46, 12],
       [12, 76, 24,  5],
       [96, 80, 27, 42],
       [71, 24, 45, 93],
       [23, 48, 47, 87]], dtype=int32)

it will pull out only which are based on the condition 

boolean array 
mask over orinal array 

In [31]:
a>50

array([[ True, False,  True,  True],
       [False, False, False, False],
       [False,  True, False, False],
       [ True,  True, False, False],
       [ True, False, False,  True],
       [False, False, False,  True]])

In [35]:
# masked 
a[a > 50]


array([93, 65, 51, 76, 96, 80, 71, 93, 87], dtype=int32)

In [36]:
# find out even numbers
a[a % 2 == 0]

array([ 4, 26, 46, 12, 12, 76, 24, 96, 80, 42, 24, 48], dtype=int32)

In [40]:
# find all numbers greater than 50 and are even
# and  this and is logical and which will not use here 
# & this will be used here 
# why --> we are working with boolean operatoro 

a[(a > 50) & (a % 2 == 0)]

array([76, 96, 80], dtype=int32)

In [43]:
# find all numbers not divisible by 7
# ~ not operator    will invert the value 
# 0-->1
# 1-->0
a[~(a % 7 == 0)]

array([93,  4, 65, 51, 17, 26, 46, 12, 12, 76, 24,  5, 96, 80, 27, 71, 24,
       45, 93, 23, 48, 47, 87], dtype=int32)

## Broadcasting

The term broadcasting describes how NumPy treats arrays with different shapes during arithmetic operations.

The smaller array is “broadcast” across the larger array so that they have compatible shapes.

# same shape


In [44]:
a = np.random.randint(1,1000,24).reshape(6,4)
b = np.random.randint(1,1000,24).reshape(6,4)

In [45]:
a

array([[214, 139, 222, 884],
       [501, 942, 645, 426],
       [338, 271, 729, 450],
       [735, 989, 698, 409],
       [ 84, 728, 755, 292],
       [886, 236, 218, 829]], dtype=int32)

In [46]:
b

array([[ 29, 428, 368, 607],
       [304, 471, 183, 998],
       [473, 309, 770, 530],
       [992,  30, 583, 745],
       [942, 946, 507, 400],
       [335, 103, 247, 369]], dtype=int32)

In [47]:
c= a + b
c

array([[ 243,  567,  590, 1491],
       [ 805, 1413,  828, 1424],
       [ 811,  580, 1499,  980],
       [1727, 1019, 1281, 1154],
       [1026, 1674, 1262,  692],
       [1221,  339,  465, 1198]], dtype=int32)

# diffrent shape 

In [51]:
a = np.arange(6).reshape(2,3)
b = np.arange(3).reshape(1,3)

a_b = a+b
a_b 

array([[0, 2, 4],
       [3, 5, 7]])

#### Broadcasting Rules

**1. Make the two arrays have the same number of dimensions.**<br>
- If the numbers of dimensions of the two arrays are different, add new dimensions with size 1 to the head of the array with the smaller dimension.<br>

**2. Make each dimension of the two arrays the same size.**<br>
- If the sizes of each dimension of the two arrays do not match, dimensions with size 1 are stretched to the size of the other array.
- If there is a dimension whose size is not 1 in either of the two arrays, it cannot be broadcasted, and an error is raised.

<img src = "https://jakevdp.github.io/PythonDataScienceHandbook/figures/02.05-broadcasting.png">

### this will work cause   

- this is due b can be   
        * b and b column number matched    
        * b can be strected 1,3 --> 2,3 ,3--> 3
        * so after finaly 4,3  (b " 4,3) can be matched 

In [52]:
# More examples

a = np.arange(12).reshape(4,3)
b = np.arange(3)

print(a)
print(b)

print(a+b)

[[ 0  1  2]
 [ 3  4  5]
 [ 6  7  8]
 [ 9 10 11]]
[0 1 2]
[[ 0  2  4]
 [ 3  5  7]
 [ 6  8 10]
 [ 9 11 13]]


not cols didnot matched due to this can't be formed

need to strech

In [53]:
# More examples

a = np.arange(12).reshape(4,4)
b = np.arange(3)

print(a)
print(b)

print(a+b)

ValueError: cannot reshape array of size 12 into shape (4,4)

In [54]:
a = np.arange(3).reshape(1,3)
b = np.arange(3).reshape(3,1)

print(a)
print(b)
 
print(a+b)

[[0 1 2]]
[[0]
 [1]
 [2]]
[[0 1 2]
 [1 2 3]
 [2 3 4]]


In [57]:
a = np.arange(3).reshape(1,3)
b = np.arange(4).reshape(4,1) 

print(a)
print(b)

print(a + b)

[[0 1 2]]
[[0]
 [1]
 [2]
 [3]]
[[0 1 2]
 [1 2 3]
 [2 3 4]
 [3 4 5]]


In [ ]:
a = np.arange(4).reshape(2,1)
 # if there  could 2--> 1 then it can be done
b = np.arange(4).reshape(4,1) 

print(a)
print(b)

print(a + b)

ValueError: cannot reshape array of size 4 into shape (3,1)